src/libostree/ostree-sysroot-deploy.c

```diff
@@ -882,46 +882,190 @@ ostree_sysroot_write_origin_file (OstreeSysroot         *sysroot,
+      in = g_unix_input_stream_new (fd, FALSE);
+      if (!ot_gio_splice_update_checksum (NULL, in, checksum, cancellable, error))
+        return FALSE;
+    }
+
+  ret_layout->bootcsum = g_strdup (g_checksum_get_string (checksum));
+  ret_layout->initialized = TRUE;
+
+  *out_layout = g_steal_pointer (&ret_layout);
+  return TRUE;
+}
+
+/* Locate kernel/initramfs in the tree; the current standard is to look in
+ * /usr/lib/modules/$kver/vmlinuz first.
+ *
+ * Originally OSTree defined kernels to be found underneath /boot
  * in the tree.  But that means when mounting /boot at runtime
  * we end up masking the content underneath, triggering a warning.
  *
  * For that reason, and also consistency with the "/usr defines the OS" model we
  * later switched to defining the in-tree kernels to be found under
- * /usr/lib/ostree-boot.
+ * /usr/lib/ostree-boot. But since then, Fedora at least switched to storing the
+ * kernel in /usr/lib/modules, which makes sense and isn't ostree-specific, so
+ * we prefer that now. However, the default Fedora layout doesn't put the
+ * initramfs there, so we need to look in /usr/lib/ostree-boot first.
  */
 static gboolean
-get_kernel_from_tree (int             deployment_dfd,
-                      int            *out_boot_dfd,
-                      char          **out_kernel_srcpath,
-                      char          **out_kernel_namever,
-                      char          **out_initramfs_srcpath,
-                      char          **out_initramfs_namever,
-                      char          **out_bootcsum,
-                      GCancellable   *cancellable,
-                      GError        **error)
+get_kernel_from_tree (int                  deployment_dfd,
+                      OstreeKernelLayout **out_layout,
+                      GCancellable        *cancellable,
+                      GError             **error)
 {
-  g_autofree char *ret_kernel_srcpath = NULL;
-  g_autofree char *ret_kernel_namever = NULL;
-  g_autofree char *ret_initramfs_srcpath = NULL;
-  g_autofree char *ret_initramfs_namever = NULL;
+  const char *legacy_paths[] = {"usr/lib/ostree-boot", "boot"};
   g_autofree char *kernel_checksum = NULL;
   g_autofree char *initramfs_checksum = NULL;
+  g_autoptr(OstreeKernelLayout) usrlib_modules_layout = NULL;
 
-  glnx_fd_close int ret_boot_dfd = glnx_opendirat_with_errno (deployment_dfd, "usr/lib/ostree-boot", TRUE);
-  if (ret_boot_dfd == -1)
+  /* First, gather from usr/lib/modules/$kver if it exists */
+  if (!get_kernel_from_tree_usrlib_modules (deployment_dfd, &usrlib_modules_layout, cancellable, error))
+    return FALSE;
+
+  /* Gather the legacy layout; if one exists, and it has an initramfs
+   * but the module layout doesn't, the legacy layout wins.  This is what happens
+   * with rpm-ostree with Fedora today, until rpm-ostree learns the new layout.
+   */
+  g_autoptr(OstreeKernelLayout) legacy_layout = g_new0 (OstreeKernelLayout, 1);
+  legacy_layout->boot_dfd = -1;
```

Comment:
*/me puts on Coverity hat*
This feels like it deserves an `_ostree_kernel_layout_new ()` to be safe, though up to you!